crates/sui-indexer/src/models_v2/epoch.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use diesel::{Insertable, Queryable};
+
+use crate::errors::IndexerError;
+use crate::schema_v2::epochs;
+use crate::types_v2::{IndexedEndOfEpochInfo, IndexedEpochInfo};
+use sui_json_rpc_types::EpochInfo;
+
+#[derive(Queryable, Insertable, Debug, Clone, Default)]
+#[diesel(table_name = epochs)]
+pub struct StoredEpochInfo {
+    pub epoch: i64,
+    pub validators: Vec<Vec<u8>>,
+    pub epoch_total_transactions: i64,
+    pub first_checkpoint_id: i64,
+    pub epoch_start_timestamp: i64,
+    // Serialized `EndOfEpochInfo`
+    pub end_of_epoch_info: Option<Vec<u8>>,
+    // Serialized `EndOfEpochData`
+    pub end_of_epoch_data: Option<Vec<u8>>,
+    pub reference_gas_price: i64,
+    pub protocol_version: i64,
+}
+
+#[derive(Queryable, Insertable, Debug, Clone, Default)]
+#[diesel(table_name = epochs)]
+pub struct StoredEndOfEpochInfo {
+    pub epoch: i64,
+    pub epoch_total_transactions: i64,
+    pub end_of_epoch_info: Vec<u8>,
+    pub end_of_epoch_data: Vec<u8>,
+
+    // For the following fields:
+    // we don't update these columns when persisting EndOfEpoch data.
+    // However if the data is partial, diesel would interpret them
+    // as Null and hence cause errors.
+    pub validators: Vec<Vec<u8>>,
+    pub first_checkpoint_id: i64,
+    pub epoch_start_timestamp: i64,
+    pub reference_gas_price: i64,
+    pub protocol_version: i64,
+}
```

Comment:
i know there are other discussions going on about this, but looking at it here it seems odd to not flatten this into the other struct.